booster/library/Booster/SMT/Runner.hs

```diff
@@ -41,64 +46,101 @@ import Booster.Log (LoggerMIO (..), logMessage)
+        , timeout = 125
+        , retryLimit = Just 3
+        , tactic = Nothing
+        }
+
 data SMTContext = SMTContext
-    { solver :: Backend.Solver
+    { options :: SMTOptions
+    , solver :: Backend.Solver
     , solverClose :: IO ()
-    , mbTranscript :: Maybe Handle
+    , mbTranscriptHandle :: Maybe Handle
+    , prelude :: [DeclareCommand]
     }
 
 ----------------------------------------
 {- TODO (later)
-- store prelude of [DeclareCommand] in context (enables hard resets)
 - error handling and retries
   - retry counter in context
-    - Reader becomes State
 - (possibly) run `get-info` on Unknown responses and enhance Unknown constructor
   - smtlib2: reason-unknown = memout | incomplete | SExpr
 -}
 
 mkContext ::
     LoggerMIO io =>
-    Maybe FilePath ->
+    SMTOptions ->
+    [DeclareCommand] ->
     io SMTContext
-mkContext transcriptPath = do
-    logMessage ("Starting new SMT solver" :: Text)
-    mbTranscript <-
-        forM transcriptPath $ \path -> do
-            logMessage $ "Transcript in file " <> pack path
-            liftIO $ do
-                h <- openFile path AppendMode
-                hSetBuffering h (BlockBuffering Nothing)
-                hSetBinaryMode h True
-                BS.hPutStrLn h "; starting solver process"
-                pure h
-    let config = Backend.defaultConfig
-    handle <- liftIO $ Backend.new config
-    solver <- liftIO $ Backend.initSolver Backend.Queuing $ Backend.toBackend handle
-    whenJust mbTranscript $ \h ->
+mkContext opts prelude = do
+    logMessage ("Starting SMT solver" :: Text)
+    (solver, handle) <- connectToSolver
+    mbTranscriptHandle <- forM opts.transcript $ \path -> do
+        logMessage $ "Transcript in file " <> pack path
+        liftIO $ do
+            h <- openFile path AppendMode
+            hSetBuffering h (BlockBuffering Nothing)
+            hSetBinaryMode h True
+            BS.hPutStrLn h "; starting solver process"
+            pure h
+    whenJust mbTranscriptHandle $ \h ->
         liftIO $ BS.hPutStrLn h "; solver initialised\n;;;;;;;;;;;;;;;;;;;;;;;"
-    logMessage ("Solver ready to use" :: Text)
     pure
         SMTContext
             { solver
             , solverClose = Backend.close handle
-            , mbTranscript
+            , mbTranscriptHandle
+            , prelude
+            , options = opts
             }
 
 closeContext :: LoggerMIO io => SMTContext -> io ()
 closeContext ctxt = do
     logMessage ("Stopping SMT solver" :: Text)
-    whenJust ctxt.mbTranscript $ \h -> liftIO $ do
-        BS.hPutStrLn h "; stopping solver\n;;;;;;;;;;;;;;;;;;;;;;;"
+    liftIO ctxt.solverClose
+
+destroyContext :: LoggerMIO io => SMTContext -> io ()
+destroyContext ctxt = do
+    logMessage ("Permanently stopping SMT solver" :: Text)
+    whenJust ctxt.mbTranscriptHandle $ \h -> liftIO $ do
+        BS.hPutStrLn h "; permanently stopping solver\n;;;;;;;;;;;;;;;;;;;;;;;"
         hClose h
     liftIO ctxt.solverClose
```

Comment:
Could you add a quick comment detailing the difference between these two? It's not immediately obvious to me when to use one or the other.